### PR TITLE
Correct scope of variables

### LINKED ISSUE #5379: Warn for unsafe variables in expressions, have consistent behavior for interpreted and compiled code
**Describe the bug**

erl_lint warns when a variable is bound in one branch of a pattern match but not another (unsafe variable).
However, there is no warning when a variable is bound in one branch of an operator expression but not another.
This is a problem because:
- Makes it so we cannot determine statically whether a variable is bound
- Leads to distinct behavior for interpreted vs. compiled code for something one would expect to be deterministic (variable binding)

**To Reproduce**

Save the following as `sample.erl`:

```erl
-module(sample).
-export([main/1]).

main(_) ->
    (begin X=1, true end)
        and
    (begin F=(fun () -> X=2 end), F(), true end),
    io:format("~p~n", [X]).
```

`escript sample.erl` warns that 'X' is unused, then uses `X`:

```
sample.erl:6:25: Warning: variable 'X' is unused
1
```

`escript -c sample.erl` (compile+run) blows up:

```
escript: exception error: no match of right hand side value 2
```

**Expected behavior**

- Warning or error for unsafe variable binding
- Consistent behavior for interpreted and compiled code

**Affected versions**

Erlang/OTP 25 [DEVELOPMENT] [erts-12.0.4] [source-5979a8621b] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns] [dtrace]

Probably earlier versions as well

Credit @rgrig @hajduakos for finding this issue

### PR DESCRIPTION
**Pull request by bjorng**
Consider this example:

    main(_) ->
        _ = (X = 42) + fun() -> X = 0 end(),
        io:format("~p\n", [X]).

When run as an `escript`, we will get different results depending
on whether we run it using `erl_eval` or using the Erlang compiler.

Here is the result using `erl_eval` (the default):

    $ escript t.erl
    42

Here is the result using the compiler:

    $ escript -c t.erl
    escript: exception error: no match of right hand side value 0

At least one of the results must be wrong. Which?

The compiler is wrong.

A binary operator (such as `+`) is allowed to evaluate its operands
in any order. Thus, code that depends on the left hand side of
a binary operator being evaluated before the right hand side is
illegal and will be rejected by the compiler. For example:

    bar(A) ->
        (X = A) + (2 * X).

Compilation will fail with the following message:

    t.erl:5:20: variable 'X' is unbound
    %    5|     (X = A) + (2 * X).
    %     |                    ^

Therefore, when we have the expression:

    (X = 42) + fun() -> X = 0 end()

the variable `X` is not bound when evaluating the right hand side
operand for `+` and the `X` bound in the fun body is another variable
named `X`, unrelated to the `X` bound on the left hand side of `+`.

This pull request corrects the compiler so that it generates code that
evalutes the example in the same way as `erl_eval` does.

Closes #5379

### CODE REVIEW

File: lib/compiler/src/v3_core.erl [hunk 1]
```diff
@@ -2147,7 +2170,155 @@ annotate_cons(A, H, T, #core{dialyzer=Dialyzer}) ->
...
+%%%
+%%% Here follows an abstract data structure to help us handle Erlang's
+%%% implicit matching that occurs when a variable is bound more than
+%%% once:
+%%%
+%%%     X = Expr1(),
+%%%     X = Expr2()
+%%%
+%%% What is implicit in Erlang, must be explicit in Core Erlang; that
+%%% is, repeated variables must eliminated and explicit matching must
```
Comment (ilya-klyuchnikov) [inline lib/compiler/src/v3_core.erl:2182]: typo: "must be eliminated"?

Comment (bjorng) [reply lib/compiler/src/v3_core.erl]: Thanks! Fixed.

File: lib/compiler/src/v3_core.erl [hunk 2]
```diff
@@ -2147,7 +2170,155 @@ annotate_cons(A, H, T, #core{dialyzer=Dialyzer}) ->
...
+%%% a set breaks down for this example:
+%%%
+%%%     (X=A) + fun() -> X = B end()
+%%%
+%%% The rewritten code would be:
+%%%
+%%%     X = A,
+%%%     Tmp = fun() ->
+%%%               X1 = B,
+%%%               if
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/v3_core.erl:2230]: > I find it rather difficult to understand why the check for the values being the same should be inside the fun. With an initial more naive rewrite, one would first have evaluation of the parts of the expression and then reconcile any bindings that have taken place in both/several branches.

I did think of that, but thought that it would be more complicated. As it turned, my original simple solution turned out to not be that simple in the end.

I think I will have a go at re-implementing my solution. It might turn out to be simpler than the current solution (or miscellaneous complications might make it impracticable or more complicated).

File: lib/compiler/test/fun_SUITE.erl [hunk 3]
```diff
@@ -296,6 +298,230 @@ unused_fun(_Config) ->
...
+    [1,100] = parallel_scopes_13(99, 100),
+    {'EXIT',{{badmatch,no},_}} = catch parallel_scopes_13(no, 100),
+    {'EXIT',{{badmatch,nope},_}} = catch parallel_scopes_13(99, nope),
+
+    ok.
+
+parallel_scopes_1a() ->
+    (begin X=1, true end
+     and
+     begin F=(fun () -> X=2 end), F(), true end) andalso X.
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/test/fun_SUITE.erl:365]: I am not sure that this is explicitly documented, but `andalso` is essentially only syntatic sugar for a `case` expression. Thus, `A andalso B` is equivalent to:

```
case A of
    true -> B;
    false -> false
end.
```

File: lib/compiler/src/v3_core.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/v3_core.erl:2230]: @lisztspace @jhogberg I have now made attempt to implement this bug fix in a cleaner way. I will try explain my thought process. This will be a long comment with examples in Core Erlang, so for anyone not comfortable with that:

**TL;DR**: The attempt didn't work out. I will soon merge this pull request in its current form.

Consider this example:
```
bar1(A, B) ->
    Z = [Y = A, Y = B],
    {Y,Z}.
```

The list elements are evaluated in parallel in any order. If the two bindings of `Y` are different, a `badmatch` exception is raised.

The compiler implements that by renaming `Y` the second time it is seen and adding a test that ensuries that the original value and new values are equal:

```
let <Y> = A
in
    case B of
      <Var1> when call 'erlang':'=:='(Var1, Y) ->
          let <Z> =
              [A, B]
          in
              {Y,Z}
      <Var2> when 'true' ->
            primop 'match_fail'({'badmatch',Var2})
    end
```

This pull request does not change this basic behavior, but ensures that the renaming and comparison does not happen in funs when a variable should not be seen.

Thinking of a different approach, I come with this Core Erlang code that actually works:

```
let <Y1, Y2> =
    <let Y = A in Y, let Y = B in Y>
in
    case <> of
      <> when call 'erlang':'=:='(Y1, Y2) ->
          let <Z> =
              [A, B]
          in
              {Y,Z}
      <> when 'true' ->
            primop 'match_fail'({'badmatch',Var2})
    end
```

Here I used the multiple values construct (`#c_values{}`) in Core Erlang. The expression inside `<` and `>` are evaluated in parallel, just as we want.

However, we get into trouble when we'll try do something slightly more complicated, such as:

```
bar2(A, B) ->
    Z = [begin Y = A, a end, begin Y = B, b end],
    {Y,Z}.
```

Since the values construct is a pure syntactic construction, it cannot be nested, so we must flatten the list of values that we will need to bind:

```
let <Y1, E1, Y2, E2> =
    <let Y = A in Y, a, let Y = B in Y, b>
in
    case <> of
      <> when call 'erlang':'=:='(Y1, Y2) ->
          let <Z> =
              [E1, E2]
          in
              {Y,Z}
      <> when 'true' ->
            primop 'match_fail'({'badmatch',Var2})
    end
```

It gets worse:

```
bar3(A, B) ->
    Z = [begin Y = A, Y + 1 end, begin Y = B, Y + 3 end],
    {Y,Z}.
```

When we try to flatten values:

```
<Y = A, Y + 1, Y = B, Y + 3>
```

it won't work, because the bindings are worked out in parallel; the binding of `Y` is not seen when `Y + 1` is evaluated.

This still simple example could be saved with some clever rewriting, but it is difficult to see how we can make it work in the general case.

So it seems that we must rename variables to get this to work in the general case. As currently implemented, the `v3_core` only renames variables when necessary. Alternatively, `v3_core` could rename all variables, keeping a mapping between the old and new names, and a list of (original) variables that are currently in scope. That would be a more massive rewrite of the `v3_core` module and the compiler would have to do more work.
